File: app/src/main/java/it/unical/mat/embasp/specializations/solver_planning_domains/desktop/SPDDesktopService.java
package it.unical.mat.embasp.specializations.solver_planning_domains.desktop;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.List;

import it.unical.mat.embasp.base.InputProgram;
import it.unical.mat.embasp.base.OptionDescriptor;
import it.unical.mat.embasp.base.Output;
import it.unical.mat.embasp.platforms.desktop.DesktopService;
import it.unical.mat.embasp.specializations.solver_planning_domains.SPDPlan;
import it.unical.mat.embasp.specializations.solver_planning_domains.SPDUtility;

public class SPDDesktopService extends DesktopService {

	private final SPDUtility spdu;

	public SPDDesktopService() {
		super("");
		spdu = new SPDUtility() {
			@Override
			protected String readFile(final String s) throws IOException {

				String everything = "";
				BufferedReader br = null;
				try {
					br = new BufferedReader(new FileReader(s));
					final StringBuilder sb = new StringBuilder();
					String line = br.readLine();

					while (line != null) {
						sb.append(line);
						sb.append(System.lineSeparator());
						line = br.readLine();
					}
					everything = sb.toString();
				} finally {
					if (br != null)
						br.close();
				}
				return everything;
			}
		};
	}

	@Override
	protected Output getOutput(final String output, final String error) {
		return new SPDPlan(output, error);
	}

	@Override
	public Output startSync(final List<InputProgram> programs, final List<OptionDescriptor> options) {

		if (programs.isEmpty())
			return getOutput("", "PDDLInputProgram not defined");

		try {
			return getOutput(spdu.postJsonToURL(spdu.createJson(programs).toString()), "");
		} catch (final Exception e) {
			return getOutput("", "Error : " + e.getMessage());
		}
	}

}

File: CSharp version/it/unical/mat/embasp/specializations/solver_planning_domains/desktop/SPDDesktopService.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace it.unical.mat.embasp.specializations.solver_planning_domains.desktop
{
    using InputProgram = it.unical.mat.embasp.@base.InputProgram;
    using OptionDescriptor = it.unical.mat.embasp.@base.OptionDescriptor;
    using Output = it.unical.mat.embasp.@base.Output;
    using DesktopService = it.unical.mat.embasp.platforms.desktop.DesktopService;

    public class SPDDesktopService : DesktopService
    {
        private readonly SPDUtility spdu;

        public SPDDesktopService() : base("") => spdu = new SPDUtilityAnonymousInnerClass(this);

        private class SPDUtilityAnonymousInnerClass : SPDUtility
        {
            private readonly SPDDesktopService outerInstance;

            public SPDUtilityAnonymousInnerClass(SPDDesktopService outerInstance) => this.outerInstance = outerInstance;

            protected internal override string ReadFile(string s)
            {
                string everything = "";
                System.IO.StreamReader br = null;
                try
                {
                    br = new System.IO.StreamReader(s);
                    StringBuilder sb = new StringBuilder();
                    string line = br.ReadLine();

                    while (!(line is null))
                    {
                        sb.Append(line);
                        sb.Append(Environment.NewLine);
                        line = br.ReadLine();
                    }
                    everything = sb.ToString();
                }
                finally
                {
                    if (br != null)
                        br.Close();
                }
                return everything;
            }
        }

        protected internal override Output GetOutput(string output, string error) => new SPDPlan(output, error);

        public override Output StartSync(IList<InputProgram> programs, IList<OptionDescriptor> options)
        {
            if (programs.Count == 0)
                return GetOutput("", "PDDLInputProgram not defined");

            try
            {
                return GetOutput(spdu.PostJsonToURL(spdu.CreateJson(programs).ToString()), "");
            }
            catch (Exception e)
            {
                return GetOutput("", "Error : " + e.Message);
            }
        }
    }
}